### ISSUE TITLE
x509 verification failure when using Name Constraints (upstream Go issue)

### ISSUE THREAD

**Issue by JonathonReinhart** [docker/docker#25031]
**TL;DR:** Docker is affected by this upstream issue: https://github.com/golang/go/issues/16347

Go's current implementation of verifying the x509 Name Constraints extension is incorrect. This causes certificates to be rejected, and eventually a failure to `docker login`.

---

I have an internal CA setup, with the certificate having an [x509 Name Constraints](https://tools.ietf.org/html/rfc5280#section-4.2.1.10) extension which allows it to only be valid for certificates for my subdomain:
```
$ openssl x509 -in example_CA.crt -text -noout
...
        Issuer: CN=Example CA
        Subject: CN=Example CA
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:0
            X509v3 Key Usage: 
                Certificate Sign, CRL Sign
            X509v3 Name Constraints: critical
                Permitted:
                  DNS:.example.com
```

Note that `DNS:.example.com` means "any host under the `example.com` domain" (e.g. `foo.example.com`).

My certificate for the Docker registry satisfies these constraints. I have confirmed a proper setup by accessing the registry API via `curl` and Google Chrome, who both are happy with the certificate.

When I attempt to `docker login`, I get the following error:

> Error response from daemon: Get https://registry.example.com/v1/users/: x509: a root or intermediate certificate is not authorized to sign in this domain

The affected version:
```
$ docker version
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:47:54 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:47:54 2016
 OS/Arch:      linux/amd64
```

This is caused by the following bug in the Go standard library: https://github.com/golang/go/issues/16347

I am opening this issue to track the status of this fix, to see that Docker is (sigh, eventually) re-built against a version of Go with this fix. It is also my intention that others encountering this error can see that it is being worked on outside of the Docker project.

**Comment by thaJeztah** [comment]
/cc @diogomonica @dmcgowan 

thanks for opening this issue @JonathonReinhart

**Comment by dmcgowan** [comment]
@JonathonReinhart while the current implementation is overly restrictive, it is still satisfying the following statement, while your proposed fix upstream no longer does (unless I am misunderstanding the context).

> DNS name restrictions are expressed as host.example.com.  Any DNS
> name that can be constructed by simply adding zero or more labels to
> the left-hand side of the name satisfies the name constraint.  For
> example, www.host.example.com would satisfy the constraint but
> host1.example.com would not.

My understanding is the work around until this is fixed up stream would be to specify the name constraint as `example.com`.

**Comment by JonathonReinhart** [comment]
@dmcgowan Yes, I'd planned on implementing that exact workaround. The conversation regarding my proposed fix is happening at https://github.com/golang/go/issues/16347.

**Comment by diogomonica** [comment]
Have we rebuilt Docker against a newer version yet?
